Sources/MapboxMaps/Style/Sources.swift

```diff
@@ -30,12 +27,32 @@ public enum SourceType: String, Codable {
         case rasterDem = "raster-dem"
         case geojson = "geojson"
         case image = "image"
-        case video = "video"
         case model = "model"
     }
+
+    /// The associated Swift struct type
+    public var sourceType: Source.Type {
+        switch self {
+        case .vector:
+            return VectorSource.self
+        case .raster:
+            return RasterSource.self
+        case .rasterDem:
+            return RasterDemSource.self
+        case .geoJson:
+            return GeoJSONSource.self
+        case .image:
+            return ImageSource.self
+        case .model:
+            return ModelSource.self
+        }
+    }
 }
 
-public protocol Source: Codable, StyleEncodable, StyleDecodable { }
+public protocol Source: Codable, StyleEncodable, StyleDecodable {
+    /// Rendering type of this source.
+    var type: SourceType { get set }
```

Comment:
If you change this to only requiring a getter, it can be satisfied by a let property. I think that makes more sense. We wouldn't want folks to be setting a type that doesn't match the struct/class.

Reply:
I think we should do this for layer too. @nishant-karajgikar would love your thoughts on this as well.

Reply:
Note that this would require a style gen change, which is fine since I already have a matching PR for this work anyways.